packages/custom-functions-metadata-plugin/src/customfunctionsplugin.ts

```diff
@@ -23,24 +27,29 @@ class CustomFunctionsMetadataPlugin {
         const inputFilePath = path.resolve(this.options.input);
         let generateResult: IGenerateResult;
 
-        compiler.hooks.compile.tap(pluginName, async () => {
+        compiler.hooks.beforeCompile.tapPromise(pluginName, async () => {
             generateResult = await generateCustomFunctionsMetadata(inputFilePath, true);
         });
 
-        compiler.hooks.emit.tap(pluginName, (compilation) => {
+        compiler.hooks.compilation.tap(pluginName, (compilation) => {
             if (generateResult.errors.length > 0) {
                 generateResult.errors.forEach((err: string) => compilation.errors.push(new WebpackError(inputFilePath + " " + err)));
             } else {
                 compilation.assets[this.options.output] = new sources.RawSource(generateResult.metadataJson);
+                if (!global.generateResults) global.generateResults = {};
+                global.generateResults[this.options.input] = generateResult;
             }
-        });
 
-        compiler.hooks.compilation.tap(pluginName, (compilation, params) => {
-            compilation.moduleTemplates.javascript.hooks.render.tap(pluginName, (source : any, module : any) => {
-                if (module._source && module._source._name.endsWith(inputFilePath)) {
-                    generateResult.associate.forEach((item: { id: string; functionName: string; }) => {
-                        module._source._value += `\nCustomFunctions.associate("${item.id}", ${item.functionName});`;
-                    });
+            NormalModule.getCompilationHooks(compilation).beforeLoaders.tap(pluginName, (_, module) => {
+                if (module.userRequest.endsWith(inputFilePath)) {
+                    module.loaders.push({
+                        loader: require.resolve("./loader.js"),
```

Comment:
It is weird that this works, given that the filename is ```loader.ts```. Could it be changed to it or not?

Reply:
It could not. It works because this only runs inside a Webpack build process, which will have this plugin installed with its transpiled files. So at runtime, `loader.ts` won't exist.